src/source/source_state.js

```diff
@@ -22,11 +22,42 @@ class SourceFeatureState {
         this.stateChanges = {};
     }
 
-    updateState(sourceLayer: string, featureId: number, state: Object) {
+    updateState(sourceLayer: string, featureId: number, newState: Object) {
         const feature = String(featureId);
         this.stateChanges[sourceLayer] = this.stateChanges[sourceLayer] || {};
         this.stateChanges[sourceLayer][feature] = this.stateChanges[sourceLayer][feature] || {};
-        extend(this.stateChanges[sourceLayer][feature], state);
+        extend(this.stateChanges[sourceLayer][feature], newState);
+
+        const deletionInQueue = this.deletedStates && this.deletedStates[sourceLayer] && this.deletedStates[sourceLayer][feature]
+        if (deletionInQueue) delete this.deletedStates[sourceLayer][feature];
+    }
+
+    removeFeatureState(sourceLayer: string, featureId: number, key: string) {
+        const feature = String(featureId);
+
+        this.deletedStates = {};
+        this.deletedStates[sourceLayer] = {};
+
+
+        if (key && featureId) {
+            this.deletedStates[sourceLayer][feature] = {};
+            this.deletedStates[sourceLayer][feature][key] = null;
+        }
+
+        else if (featureId) {
+            this.deletedStates[sourceLayer][feature] = {};
+            for (key in this.state[sourceLayer][feature]) this.deletedStates[sourceLayer][feature][key] = null;
+        }
+
+        else  {
+            this.deletedStates[sourceLayer] = {};
+            for (feature in this.state[sourceLayer]) {
+                this.deletedStates[sourceLayer][feature] = {};
+                for (key in this.state[sourceLayer][feature]) {
+                    this.deletedStates[sourceLayer][feature][key] = null;
```

Comment:
Is it necessary to iterate over all features/keys here? Can't presence of an empty `this.deleteStates[sourceLayer]` object be used to indicate that all items should be deleted? Similarly for all items for a specific feature id ?

Reply:
🤔it appears necessary, because in the coalesce step `layerStates` needs to iterate over features from `deletedStates` to populate itself: https://github.com/mapbox/mapbox-gl-js/blob/dbbf41cf26acd1903e8c39f313fd1979f37924b7/src/source/source_state.js#L98

In turn, layerStates is used to extend `currentChanges`

Reply:
If I understand correctly, this whole workflow works by an "absence of evidence is not evidence of absence" philosophy, which necessitates the explicit `null`s